src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -4603,16 +4625,47 @@ public EventLevel Level
             }
         }
 
-#region private
+        /// <summary>
+        /// Gets the identifier for the OS thread that wrote the event.
+        /// </summary>
+        public int OSThreadId
```

Comment:
`int32` OS thread id is Windows specific concept. It does not exist on Unix. The CoreCLR Win32 PAL fakes the implementation to return something, but that implementation is less than ideal (e.g. it can return same thread ID for multiple threads in the given process).

Reply:
Related:

https://github.com/dotnet/coreclr/issues/889 is related to this.
https://github.com/dotnet/coreclr/issues/13037#issuecomment-318137113

IIRC, the fake Win32 threadID has been a problem for the debugger and it required some creative workarounds there.

Reply:
I think it depends what the expected use of this property is. This PR does not say what we expect this to be used for. Since these are new public APIs, it would be nice to use the [process and template for adding public APIs](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) so that we have information like this captured and taken into account.

I can think about two options:

- Return the actual OS thread ID. It means that it has to be 64-bit number for portability to non-Windows OSes. The implementation should return the actual OS thread ID (i.e. what `pthread_threadid_np` returns), not the Win32 PAL fake thread ID.

- Return the managed thread ID. Managed thread IDs are 32-bit, managed-code specific concept.

Reply:
I can give information on what this OSThreadID will be used for, as this was meant to mimic similar APIs in the Windows ETW interface.     The OSThreadID's main use is as a correlator.  It allows you to 'link' events that are related (in this case that they happened on the same thread).     It is used pervasively.

Now either of the options above can be used for that purpose, so that does not help.   It is actually helpful if the IDs are NOT REUSED AGRESSIVELY, since if they are reused commonly, the parsing logic has to go to some trouble to basically make they more globally unique again (otherwise you get false correlations if a thread dies and another thread reuses the ID).     Now arguably that reuse problem is there for OSThreadID, but in practice most code simply ignores the problem (which works fine for OSThreadID most of the time).  

The other value of using the OSThreadID is that it is certainly possible that the data will be persisted and correlated with data that comes from other tools (e.g. the Linux Perf tool), that does not have the concept of the managed thread id.    This creates unnecessary pain trying to corelate things.   (And this is another reason you want the 'true' OSThreadID now any one that was morphed by our PAL). 

I think given this, the guidance should be simply to expand the OSThreadID to be a 64 bit number (presumably a long).

Reply:
@vancem If this is going to be used for doing similar analysis to ETW, would it make more sense to fall back to the [`UniqueProcessKey`](https://github.com/Microsoft/perfview/blob/dcece4936e6c8c60073eddd2213b00bca83aab69/src/TraceEvent/Parsers/KernelTraceEventParser.cs#L3837) concept which is `size_t` on Windows anyways? This is what various "other" tools I am familiar with use and might be useful for them in the future as well 😄

Reply:
@AaronRobinsonMSFT - You will note UniqueProcessKey is not what is in every event, but rather in the Process Event, which is then correlated to the (less globally unique) process key.   There is a reason for this (part legacy, but mostly that uniqueness comes at a cost, and you want your system to be pay-for-play (things that don't need global uniqueness don't pay for it)).    

As a practical mater, I don't think we have a similarly globally unique  ID for threads, and we probably don't need it (if thread IDs are unique within a process, that would be enough), and because we want to operate on multiple OSes we likely want to stick with things that are available 'everywhere'.   

I think are guidance currently si that we should use the true OS thread ID, and thus make the value 64 bit.   How woudl your insight change that suggestion?

Reply:
Shoot - sorry for the noise. When you mentioned the 64-bit size, I remembered a unique key of `size_t` and completely ignored the fact that this was related to threads _not_ processes.

My basic thought here was an attempt to keep the underlying data model consistent between the two platforms (i.e. type size on Windows and type size on Non-Windows). Since thread ID in the eventing system is already 32-bit I was thinking of using a different but present ID mechanism in ETW that is already 64-bit or `size_t`.

In this case, there isn't one because the only thread ID-esque concept on an ETW event is 32-bit.

Reply:
I have filed https://github.com/dotnet/corefx/issues/31401 to review the new APIs.